### PR TITLE
Fix 18068: Updates merge_asof error, now outputs datatypes

### LINKED ISSUE #18068: ERR: better error message on merging for incompat types
would be nice to have the types outputted in the error message

```
In [23]:         left = pd.DataFrame({'a': [1, 5, 10],
    ...:                              'left_val': ['a', 'b', 'c']})
    ...:         right = pd.DataFrame({'a': [1, 2, 3, 6, 7],
    ...:                               'right_val': [1, 2, 3, 6, 7]})
    ...: 
    ...: 

In [24]: pd.merge_asof(left, right, on='a')
Out[24]: 
    a left_val  right_val
0   1        a          1
1   5        b          3
2  10        c          7

In [25]: pd.merge_asof(left, right.assign(a=right.a.astype('i4')), on='a')
MergeError: incompatible merge keys, must be the same type
```

### PR DESCRIPTION
**Pull request by manrajgrover**
- [ ] closes #18068
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/core/reshape/merge.py [hunk 1]
```diff
@@ -1255,8 +1255,9 @@ def _get_merge_keys(self):
         # validate index types are the same
         for lk, rk in zip(left_join_keys, right_join_keys):
             if not is_dtype_equal(lk.dtype, rk.dtype):
-                raise MergeError("incompatible merge keys, "
-                                 "must be the same type")
+                raise MergeError("incompatible merge keys {lkdtype} and "
+                                 "{rkdtype}, must be the same type"
+                                 .format(lkdtype=lk.dtype, rkdtype=rk.dtype))
```
Comment (jreback) [inline pandas/core/reshape/merge.py]: might also add an enumerate here to make it even more clear, e.g.

```
for i, (lk, rk) in enumerate(zip(......))):
    .....
     raise MergeError("incompatible merge key [{i} {lkdtype} and "
```

**Review by jreback — CHANGES_REQUESTED** [review]
can you add a whatsnew note for 0.21.1 as well.

**Comment by ManrajGrover** [comment]
@jreback Made the requested changes. I have a question though. How should I check whether release notes messages will successfully get parsed? Is there a build command to check that?

File: pandas/core/reshape/merge.py [hunk 2]
```diff
@@ -1253,10 +1253,12 @@ def _get_merge_keys(self):
          join_names) = super(_AsOfMerge, self)._get_merge_keys()
 
         # validate index types are the same
-        for lk, rk in zip(left_join_keys, right_join_keys):
+        for i, (lk, rk) in enumerate(zip(left_join_keys, right_join_keys)):
             if not is_dtype_equal(lk.dtype, rk.dtype):
-                raise MergeError("incompatible merge keys, "
-                                 "must be the same type")
+                raise MergeError("incompatible merge keys [{i} {lkdtype} and "
```
Comment (jreback) [inline pandas/core/reshape/merge.py]: [{i}] other wise this is confusing

**Review by jreback — CHANGES_REQUESTED** [review]
you can use tm.assert_raises_regex to assert the errors messages

File: doc/source/whatsnew/v0.21.1.txt [hunk 3]
```diff
@@ -102,7 +102,7 @@ Sparse
 Reshaping
 ^^^^^^^^^
 
--
+- Error message in ``pd.merge_asof()`` for key datatype mismatch now includes datatype of left and right key (:issue:`18068`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.1.txt]: can you use `` :func:`merge_asof` `` so we get the link

**Review by jreback — CHANGES_REQUESTED** [review]
can you add a test which asserts this error message (e.g. use the case from the original issue)

File: doc/source/whatsnew/v0.21.1.txt (hunk 3 above)
Comment (ManrajGrover) [reply doc/source/whatsnew/v0.21.1.txt]: @jreback Is there a documentation for syntax to be used for release notes? Also, how can I view rendered version of it?

Making the required change. Also, will add test for the same.

Comment (jreback) [reply doc/source/whatsnew/v0.21.1.txt]: no this is fine

Comment (jreback) [reply doc/source/whatsnew/v0.21.1.txt]: documentation is look at similar notes!

Comment (jreback) [reply doc/source/whatsnew/v0.21.1.txt]: you would have to build the docs to render it

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @ManrajGrover

**Pull request merged**